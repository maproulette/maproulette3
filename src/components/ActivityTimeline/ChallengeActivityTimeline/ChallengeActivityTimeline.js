import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { injectIntl } from 'react-intl'
import classNames from 'classnames'
import _map from 'lodash/map'
import _compact from 'lodash/compact'
import _groupBy from 'lodash/groupBy'
import _toPairs from 'lodash/toPairs'
import _reverse from 'lodash/reverse'
import _sortBy from 'lodash/sortBy'
import _take from 'lodash/take'
import _isFinite from 'lodash/isFinite'
import parse from 'date-fns/parse'
import { statusLabels,
         keysByStatus }
       from '../../../services/Task/TaskStatus/TaskStatus'
import '../ActivityTimeline.scss'

/**
 * ChallengeActivityTimeline displays recent challenge activity from the given
 * challenge in the form of a Bulma timeline, with the most recent activity
 * shown at the top. Activity is grouped together by day.
 *
 * @see https://wikiki.github.io/components/timeline/
 *
 * @author [Neil Rotstan](https://github.com/nrotstan)
 */
export class ChallengeActivityTimeline extends Component {
  render() {
    const localizedStatusLabels = statusLabels(this.props.intl)

    // Each status will receive its own entry for a given date (with a count of
    // zero if there is nothing on that day for that status). We group them
    // together by date, then sort in descending order by date so the most
    // recent activity will show up first.
    const groupedByDate = _toPairs(_groupBy(this.props.activity, 'date'))
    const latestEntries = _reverse(
      _sortBy(groupedByDate, (value) => parse(value[0]).getTime())
    )

    // Build timeline entries for each day, showing each (non-zero) type of
    // activity with a badge displaying the count.
    let timelineItems = _compact(_map(latestEntries, entriesByDate => {
      const isoDate = entriesByDate[0]
      const statusEntries = entriesByDate[1]

      const formattedDate = this.props.intl.formatDate(parse(isoDate),
                                                       {month: 'long', day: 'numeric'})
      const statuses = _compact(_map(statusEntries, entry => {
        if (entry.count === 0) {
          return null
        }

        return (
          <p key={`${isoDate}-${entry.status}`}
             className="timeline-item__activity-entry">
            <span className={classNames("badge", {inverted: this.props.invertBadges})}
                  data-badge={entry.count}>
              {localizedStatusLabels[keysByStatus[entry.status]]}
            </span>
          </p>
        )
      }))

      return statuses.length === 0 ? null : (
        <div key={isoDate} className="timeline-item">
          <div className="timeline-marker"></div>
          <div className="timeline-content">
            <p className="heading">{formattedDate}</p>
            {statuses}
          </div>
        </div>
      )
    }))

    // If a limit on timeline entries has been given, honor it.
    if (_isFinite(this.props.maxEntries)) {
      timelineItems = _take(timelineItems, this.props.maxEntries)
    }

    // If there is no reportable activity, show that on the timeline.
    if (timelineItems.length === 0) {
      timelineItems.push(
        <div key="no-activity" className="timeline-item no-activity">
          <div className="timeline-content">
            <p className="heading">No Recent Activity</p>
          </div>
        </div>
      )
    }

    return (
      <div className="timeline activity-timeline">
        <header className="timeline-header">
          <span className="tag is-medium">Latest</span>
        </header>

        {timelineItems}

        <div className="timeline-header">
          <span className="tag is-medium">End</span>
        </div>
      </div>
    )
  }
}

ChallengeActivityTimeline.propTypes = {
  /** The activity to display on the timeline */
  activity: PropTypes.arrayOf(PropTypes.shape({
    date: PropTypes.string.isRequired,
    status: PropTypes.number.isRequired,
    count: PropTypes.number.isRequired,
    statusName: PropTypes.string,
  })),
  invertBadges: PropTypes.bool,
}

ChallengeActivityTimeline.defaultProps = {
  activity: [],
  invertBadges: true,
}

export default injectIntl(ChallengeActivityTimeline)
