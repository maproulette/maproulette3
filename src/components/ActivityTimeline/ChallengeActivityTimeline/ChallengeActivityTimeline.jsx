import { parseISO } from "date-fns";
import _compact from "lodash/compact";
import _groupBy from "lodash/groupBy";
import _map from "lodash/map";
import _reverse from "lodash/reverse";
import _sortBy from "lodash/sortBy";
import _take from "lodash/take";
import _toPairs from "lodash/toPairs";
import PropTypes from "prop-types";
import { Component } from "react";
import { injectIntl } from "react-intl";
import { keysByStatus, statusLabels } from "../../../services/Task/TaskStatus/TaskStatus";
import "../ActivityTimeline.scss";

/**
 * ChallengeActivityTimeline displays recent challenge activity from the given
 * challenge, with the most recent activity
 * shown at the top. Activity is grouped together by day.
 *
 * @see https://wikiki.github.io/components/timeline/
 *
 * @author [Neil Rotstan](https://github.com/nrotstan)
 */
export class ChallengeActivityTimeline extends Component {
  render() {
    const localizedStatusLabels = statusLabels(this.props.intl);

    // Each status will receive its own entry for a given date (with a count of
    // zero if there is nothing on that day for that status). We group them
    // together by date, then sort in descending order by date so the most
    // recent activity will show up first.
    const groupedByDate = _toPairs(_groupBy(this.props.activity, "date"));
    const latestEntries = _reverse(_sortBy(groupedByDate, (value) => parseISO(value[0]).getTime()));

    // Build timeline entries for each day, showing each (non-zero) type of
    // activity with a badge displaying the count.
    let timelineItems = _compact(
      _map(latestEntries, (entriesByDate) => {
        const isoDate = entriesByDate[0];
        const statusEntries = entriesByDate[1];

        const formattedDate = this.props.intl.formatDate(parseISO(isoDate), {
          month: "long",
          day: "numeric",
        });
        const statuses = _compact(
          _map(statusEntries, (entry) => {
            if (entry.count === 0) {
              return null;
            }

            return (
              <li key={`${isoDate}-${entry.status}`}>
                <span className="mr-badge">{entry.count}</span>{" "}
                {localizedStatusLabels[keysByStatus[entry.status]]}
              </li>
            );
          }),
        );

        return statuses.length === 0 ? null : (
          <li key={isoDate} className="mr-timeline__period">
            <h3 className="mr-timeline__header">{formattedDate}</h3>
            <ol className="mr-timeline__activity">{statuses}</ol>
          </li>
        );
      }),
    );

    // If a limit on timeline entries has been given, honor it.
    if (Number.isFinite(this.props.maxEntries)) {
      timelineItems = _take(timelineItems, this.props.maxEntries);
    }

    // If there is no reportable activity, show that on the timeline.
    if (timelineItems.length === 0) {
      timelineItems.push(
        <div key="no-activity" className="timeline-item no-activity">
          <div className="timeline-content">
            <p className="heading">No Recent Activity</p>
          </div>
        </div>,
      );
    }

    return <ol className="mr-timeline dark-mode">{timelineItems}</ol>;
  }
}

ChallengeActivityTimeline.propTypes = {
  /** The activity to display on the timeline */
  activity: PropTypes.arrayOf(
    PropTypes.shape({
      date: PropTypes.string.isRequired,
      status: PropTypes.number.isRequired,
      count: PropTypes.number.isRequired,
      statusName: PropTypes.string,
    }),
  ),
  invertBadges: PropTypes.bool,
};

ChallengeActivityTimeline.defaultProps = {
  activity: [],
  invertBadges: true,
};

export default injectIntl(ChallengeActivityTimeline);
