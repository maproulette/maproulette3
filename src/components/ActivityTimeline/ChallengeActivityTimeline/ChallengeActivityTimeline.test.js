import "@testing-library/jest-dom";
import { fireEvent } from "@testing-library/react";
import { ChallengeActivityTimeline } from "./ChallengeActivityTimeline";

describe("ChallengeActivityTimeline", () => {
  it("renders ChallengeActivityTimeline successfully", () => {
    const { container } = global.withProvider(
      <ChallengeActivityTimeline intl={{ formatMessage: () => null }}/>
    );

    expect(container.firstChild.classList["0"]).toBe("mr-timeline");
  });

  it("renders ChallengeActivityTimeline successfully", () => {
    const { getByText } = global.withProvider(
      <ChallengeActivityTimeline intl={{ formatMessage: () => null, formatDate: () => null }} activity={[{
        count: 10,
        date: "2022-08-03T00:00:00",
        status: 3,
        statusName: "Skipped",
      }]}/>
    );

    const text = getByText("10");
    expect(text).toBeInTheDocument();
  });
});