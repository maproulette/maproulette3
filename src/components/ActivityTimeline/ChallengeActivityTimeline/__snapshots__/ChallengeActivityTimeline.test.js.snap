// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it indicates on the timeline if there is no activity 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <ChallengeActivityTimeline
    activity={Array []}
    intl={
      Object {
        "formatDate": [MockFunction],
        "formatMessage": [MockFunction] {
          "calls": Array [
            Array [
              Object {
                "defaultMessage": "Created",
                "id": "Task.status.created",
              },
            ],
            Array [
              Object {
                "defaultMessage": "Fixed",
                "id": "Task.status.fixed",
              },
            ],
            Array [
              Object {
                "defaultMessage": "Not an Issue",
                "id": "Task.status.falsePositive",
              },
            ],
            Array [
              Object {
                "defaultMessage": "Skipped",
                "id": "Task.status.skipped",
              },
            ],
            Array [
              Object {
                "defaultMessage": "Deleted",
                "id": "Task.status.deleted",
              },
            ],
            Array [
              Object {
                "defaultMessage": "Already Fixed",
                "id": "Task.status.alreadyFixed",
              },
            ],
            Array [
              Object {
                "defaultMessage": "Too Hard",
                "id": "Task.status.tooHard",
              },
            ],
          ],
          "results": Array [
            Object {
              "isThrow": false,
              "value": undefined,
            },
            Object {
              "isThrow": false,
              "value": undefined,
            },
            Object {
              "isThrow": false,
              "value": undefined,
            },
            Object {
              "isThrow": false,
              "value": undefined,
            },
            Object {
              "isThrow": false,
              "value": undefined,
            },
            Object {
              "isThrow": false,
              "value": undefined,
            },
            Object {
              "isThrow": false,
              "value": undefined,
            },
          ],
        },
      }
    }
    invertBadges={true}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <header
          className="timeline-header"
        >
          <span
            className="tag is-medium"
          >
            Latest
          </span>
        </header>,
        Array [
          <div
            className="timeline-item no-activity"
          >
            <div
              className="timeline-content"
            >
              <p
                className="heading"
              >
                No Recent Activity
              </p>
            </div>
          </div>,
        ],
        <div
          className="timeline-header"
        >
          <span
            className="tag is-medium"
          >
            End
          </span>
        </div>,
      ],
      "className": "timeline activity-timeline",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <span
            className="tag is-medium"
          >
            Latest
          </span>,
          "className": "timeline-header",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Latest",
            "className": "tag is-medium",
          },
          "ref": null,
          "rendered": "Latest",
          "type": "span",
        },
        "type": "header",
      },
      Object {
        "instance": null,
        "key": "no-activity",
        "nodeType": "host",
        "props": Object {
          "children": <div
            className="timeline-content"
          >
            <p
              className="heading"
            >
              No Recent Activity
            </p>
          </div>,
          "className": "timeline-item no-activity",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <p
              className="heading"
            >
              No Recent Activity
            </p>,
            "className": "timeline-content",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "No Recent Activity",
              "className": "heading",
            },
            "ref": null,
            "rendered": "No Recent Activity",
            "type": "p",
          },
          "type": "div",
        },
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <span
            className="tag is-medium"
          >
            End
          </span>,
          "className": "timeline-header",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "End",
            "className": "tag is-medium",
          },
          "ref": null,
          "rendered": "End",
          "type": "span",
        },
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <header
            className="timeline-header"
          >
            <span
              className="tag is-medium"
            >
              Latest
            </span>
          </header>,
          Array [
            <div
              className="timeline-item no-activity"
            >
              <div
                className="timeline-content"
              >
                <p
                  className="heading"
                >
                  No Recent Activity
                </p>
              </div>
            </div>,
          ],
          <div
            className="timeline-header"
          >
            <span
              className="tag is-medium"
            >
              End
            </span>
          </div>,
        ],
        "className": "timeline activity-timeline",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <span
              className="tag is-medium"
            >
              Latest
            </span>,
            "className": "timeline-header",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Latest",
              "className": "tag is-medium",
            },
            "ref": null,
            "rendered": "Latest",
            "type": "span",
          },
          "type": "header",
        },
        Object {
          "instance": null,
          "key": "no-activity",
          "nodeType": "host",
          "props": Object {
            "children": <div
              className="timeline-content"
            >
              <p
                className="heading"
              >
                No Recent Activity
              </p>
            </div>,
            "className": "timeline-item no-activity",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <p
                className="heading"
              >
                No Recent Activity
              </p>,
              "className": "timeline-content",
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "No Recent Activity",
                "className": "heading",
              },
              "ref": null,
              "rendered": "No Recent Activity",
              "type": "p",
            },
            "type": "div",
          },
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <span
              className="tag is-medium"
            >
              End
            </span>,
            "className": "timeline-header",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "End",
              "className": "tag is-medium",
            },
            "ref": null,
            "rendered": "End",
            "type": "span",
          },
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;

exports[`it renders a timeline of the activity 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <ChallengeActivityTimeline
    activity={
      Array [
        Object {
          "count": 2,
          "date": "2017-11-22T00:00:00",
          "status": 1,
          "statusName": "Fixed",
        },
        Object {
          "count": 0,
          "date": "2017-11-22T00:00:00",
          "status": 3,
          "statusName": "Skipped",
        },
        Object {
          "count": 1,
          "date": "2017-11-23T00:00:00",
          "status": 1,
          "statusName": "Fixed",
        },
      ]
    }
    intl={
      Object {
        "formatDate": [MockFunction] {
          "calls": Array [
            Array [
              2017-11-23T08:00:00.000Z,
              Object {
                "day": "numeric",
                "month": "long",
              },
            ],
            Array [
              2017-11-22T08:00:00.000Z,
              Object {
                "day": "numeric",
                "month": "long",
              },
            ],
          ],
          "results": Array [
            Object {
              "isThrow": false,
              "value": undefined,
            },
            Object {
              "isThrow": false,
              "value": undefined,
            },
          ],
        },
        "formatMessage": [MockFunction] {
          "calls": Array [
            Array [
              Object {
                "defaultMessage": "Created",
                "id": "Task.status.created",
              },
            ],
            Array [
              Object {
                "defaultMessage": "Fixed",
                "id": "Task.status.fixed",
              },
            ],
            Array [
              Object {
                "defaultMessage": "Not an Issue",
                "id": "Task.status.falsePositive",
              },
            ],
            Array [
              Object {
                "defaultMessage": "Skipped",
                "id": "Task.status.skipped",
              },
            ],
            Array [
              Object {
                "defaultMessage": "Deleted",
                "id": "Task.status.deleted",
              },
            ],
            Array [
              Object {
                "defaultMessage": "Already Fixed",
                "id": "Task.status.alreadyFixed",
              },
            ],
            Array [
              Object {
                "defaultMessage": "Too Hard",
                "id": "Task.status.tooHard",
              },
            ],
          ],
          "results": Array [
            Object {
              "isThrow": false,
              "value": undefined,
            },
            Object {
              "isThrow": false,
              "value": undefined,
            },
            Object {
              "isThrow": false,
              "value": undefined,
            },
            Object {
              "isThrow": false,
              "value": undefined,
            },
            Object {
              "isThrow": false,
              "value": undefined,
            },
            Object {
              "isThrow": false,
              "value": undefined,
            },
            Object {
              "isThrow": false,
              "value": undefined,
            },
          ],
        },
      }
    }
    invertBadges={true}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <header
          className="timeline-header"
        >
          <span
            className="tag is-medium"
          >
            Latest
          </span>
        </header>,
        Array [
          <div
            className="timeline-item"
          >
            <div
              className="timeline-marker"
            />
            <div
              className="timeline-content"
            >
              <p
                className="heading"
              />
              <p
                className="timeline-item__activity-entry"
              >
                <span
                  className="badge inverted"
                  data-badge={1}
                />
              </p>
            </div>
          </div>,
          <div
            className="timeline-item"
          >
            <div
              className="timeline-marker"
            />
            <div
              className="timeline-content"
            >
              <p
                className="heading"
              />
              <p
                className="timeline-item__activity-entry"
              >
                <span
                  className="badge inverted"
                  data-badge={2}
                />
              </p>
            </div>
          </div>,
        ],
        <div
          className="timeline-header"
        >
          <span
            className="tag is-medium"
          >
            End
          </span>
        </div>,
      ],
      "className": "timeline activity-timeline",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <span
            className="tag is-medium"
          >
            Latest
          </span>,
          "className": "timeline-header",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Latest",
            "className": "tag is-medium",
          },
          "ref": null,
          "rendered": "Latest",
          "type": "span",
        },
        "type": "header",
      },
      Object {
        "instance": null,
        "key": "2017-11-23T00:00:00",
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <div
              className="timeline-marker"
            />,
            <div
              className="timeline-content"
            >
              <p
                className="heading"
              />
              <p
                className="timeline-item__activity-entry"
              >
                <span
                  className="badge inverted"
                  data-badge={1}
                />
              </p>
            </div>,
          ],
          "className": "timeline-item",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "className": "timeline-marker",
            },
            "ref": null,
            "rendered": null,
            "type": "div",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <p
                  className="heading"
                />,
                Array [
                  <p
                    className="timeline-item__activity-entry"
                  >
                    <span
                      className="badge inverted"
                      data-badge={1}
                    />
                  </p>,
                ],
              ],
              "className": "timeline-content",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": undefined,
                  "className": "heading",
                },
                "ref": null,
                "rendered": null,
                "type": "p",
              },
              Object {
                "instance": null,
                "key": "2017-11-23T00:00:00-1",
                "nodeType": "host",
                "props": Object {
                  "children": <span
                    className="badge inverted"
                    data-badge={1}
                  />,
                  "className": "timeline-item__activity-entry",
                },
                "ref": null,
                "rendered": Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": undefined,
                    "className": "badge inverted",
                    "data-badge": 1,
                  },
                  "ref": null,
                  "rendered": null,
                  "type": "span",
                },
                "type": "p",
              },
            ],
            "type": "div",
          },
        ],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": "2017-11-22T00:00:00",
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <div
              className="timeline-marker"
            />,
            <div
              className="timeline-content"
            >
              <p
                className="heading"
              />
              <p
                className="timeline-item__activity-entry"
              >
                <span
                  className="badge inverted"
                  data-badge={2}
                />
              </p>
            </div>,
          ],
          "className": "timeline-item",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "className": "timeline-marker",
            },
            "ref": null,
            "rendered": null,
            "type": "div",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <p
                  className="heading"
                />,
                Array [
                  <p
                    className="timeline-item__activity-entry"
                  >
                    <span
                      className="badge inverted"
                      data-badge={2}
                    />
                  </p>,
                ],
              ],
              "className": "timeline-content",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": undefined,
                  "className": "heading",
                },
                "ref": null,
                "rendered": null,
                "type": "p",
              },
              Object {
                "instance": null,
                "key": "2017-11-22T00:00:00-1",
                "nodeType": "host",
                "props": Object {
                  "children": <span
                    className="badge inverted"
                    data-badge={2}
                  />,
                  "className": "timeline-item__activity-entry",
                },
                "ref": null,
                "rendered": Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": undefined,
                    "className": "badge inverted",
                    "data-badge": 2,
                  },
                  "ref": null,
                  "rendered": null,
                  "type": "span",
                },
                "type": "p",
              },
            ],
            "type": "div",
          },
        ],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <span
            className="tag is-medium"
          >
            End
          </span>,
          "className": "timeline-header",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "End",
            "className": "tag is-medium",
          },
          "ref": null,
          "rendered": "End",
          "type": "span",
        },
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <header
            className="timeline-header"
          >
            <span
              className="tag is-medium"
            >
              Latest
            </span>
          </header>,
          Array [
            <div
              className="timeline-item"
            >
              <div
                className="timeline-marker"
              />
              <div
                className="timeline-content"
              >
                <p
                  className="heading"
                />
                <p
                  className="timeline-item__activity-entry"
                >
                  <span
                    className="badge inverted"
                    data-badge={1}
                  />
                </p>
              </div>
            </div>,
            <div
              className="timeline-item"
            >
              <div
                className="timeline-marker"
              />
              <div
                className="timeline-content"
              >
                <p
                  className="heading"
                />
                <p
                  className="timeline-item__activity-entry"
                >
                  <span
                    className="badge inverted"
                    data-badge={2}
                  />
                </p>
              </div>
            </div>,
          ],
          <div
            className="timeline-header"
          >
            <span
              className="tag is-medium"
            >
              End
            </span>
          </div>,
        ],
        "className": "timeline activity-timeline",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <span
              className="tag is-medium"
            >
              Latest
            </span>,
            "className": "timeline-header",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Latest",
              "className": "tag is-medium",
            },
            "ref": null,
            "rendered": "Latest",
            "type": "span",
          },
          "type": "header",
        },
        Object {
          "instance": null,
          "key": "2017-11-23T00:00:00",
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <div
                className="timeline-marker"
              />,
              <div
                className="timeline-content"
              >
                <p
                  className="heading"
                />
                <p
                  className="timeline-item__activity-entry"
                >
                  <span
                    className="badge inverted"
                    data-badge={1}
                  />
                </p>
              </div>,
            ],
            "className": "timeline-item",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "className": "timeline-marker",
              },
              "ref": null,
              "rendered": null,
              "type": "div",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  <p
                    className="heading"
                  />,
                  Array [
                    <p
                      className="timeline-item__activity-entry"
                    >
                      <span
                        className="badge inverted"
                        data-badge={1}
                      />
                    </p>,
                  ],
                ],
                "className": "timeline-content",
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": undefined,
                    "className": "heading",
                  },
                  "ref": null,
                  "rendered": null,
                  "type": "p",
                },
                Object {
                  "instance": null,
                  "key": "2017-11-23T00:00:00-1",
                  "nodeType": "host",
                  "props": Object {
                    "children": <span
                      className="badge inverted"
                      data-badge={1}
                    />,
                    "className": "timeline-item__activity-entry",
                  },
                  "ref": null,
                  "rendered": Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "host",
                    "props": Object {
                      "children": undefined,
                      "className": "badge inverted",
                      "data-badge": 1,
                    },
                    "ref": null,
                    "rendered": null,
                    "type": "span",
                  },
                  "type": "p",
                },
              ],
              "type": "div",
            },
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": "2017-11-22T00:00:00",
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <div
                className="timeline-marker"
              />,
              <div
                className="timeline-content"
              >
                <p
                  className="heading"
                />
                <p
                  className="timeline-item__activity-entry"
                >
                  <span
                    className="badge inverted"
                    data-badge={2}
                  />
                </p>
              </div>,
            ],
            "className": "timeline-item",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "className": "timeline-marker",
              },
              "ref": null,
              "rendered": null,
              "type": "div",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  <p
                    className="heading"
                  />,
                  Array [
                    <p
                      className="timeline-item__activity-entry"
                    >
                      <span
                        className="badge inverted"
                        data-badge={2}
                      />
                    </p>,
                  ],
                ],
                "className": "timeline-content",
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": undefined,
                    "className": "heading",
                  },
                  "ref": null,
                  "rendered": null,
                  "type": "p",
                },
                Object {
                  "instance": null,
                  "key": "2017-11-22T00:00:00-1",
                  "nodeType": "host",
                  "props": Object {
                    "children": <span
                      className="badge inverted"
                      data-badge={2}
                    />,
                    "className": "timeline-item__activity-entry",
                  },
                  "ref": null,
                  "rendered": Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "host",
                    "props": Object {
                      "children": undefined,
                      "className": "badge inverted",
                      "data-badge": 2,
                    },
                    "ref": null,
                    "rendered": null,
                    "type": "span",
                  },
                  "type": "p",
                },
              ],
              "type": "div",
            },
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <span
              className="tag is-medium"
            >
              End
            </span>,
            "className": "timeline-header",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "End",
              "className": "tag is-medium",
            },
            "ref": null,
            "rendered": "End",
            "type": "span",
          },
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
